### PR TITLE
Add check for insertionOrder and unique for drift parity

### PR DESCRIPTION
**Pull request by anshikg**
*Issue #, if available:* https://github.com/aws-cloudformation/cloudformation-cli/issues/580

*Description of changes:* This CR has a change for insertionOrder metadata check. It is added to ensure drift parity. This metadata if set to false ensures order does not matters

*Test*
* unit test
* ran contract test
```
(env) 3c22fbb25c23:aws-cloudwatch-alarm anshikg$ cfn test --enforce-timeout 120
=================================================================================================================== test session starts ====================================================================================================================
platform darwin -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /Users/anshikg/workspace/cloudformation-cli/env/bin/python3
cachedir: .pytest_cache
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/anshikg/workspace/rpdk/aws-cloudformation-resource-providers-cloudwatch/aws-cloudwatch-alarm/.hypothesis/examples')
rootdir: /Users/anshikg/workspace/rpdk/aws-cloudformation-resource-providers-cloudwatch/aws-cloudwatch-alarm, configfile: ../../../../../../private/var/folders/zh/rhw_046j7bq4d88hjn671hqjy323j8/T/pytest_zitq3qml.ini
plugins: random-order-1.0.4, localserver-0.5.0, cov-2.10.0, hypothesis-5.19.3, black-0.3.10
collected 16 items                                                                                                                                                                                                                                         

handler_create.py::contract_create_delete PASSED                                                                                                                                                                                                     [  6%]
handler_create.py::contract_invalid_create PASSED                                                                                                                                                                                                    [ 12%]
handler_create.py::contract_create_duplicate PASSED                                                                                                                                                                                                  [ 18%]
handler_create.py::contract_create_read_success PASSED                                                                                                                                                                                               [ 25%]
handler_create.py::contract_create_list_success PASSED                                                                                                                                                                                               [ 31%]
handler_delete.py::contract_delete_read PASSED                                                                                                                                                                                                       [ 37%]
handler_delete.py::contract_delete_list PASSED                                                                                                                                                                                                       [ 43%]
handler_delete.py::contract_delete_update PASSED                                                                                                                                                                                                     [ 50%]
handler_delete.py::contract_delete_delete PASSED                                                                                                                                                                                                     [ 56%]
handler_delete.py::contract_delete_create PASSED                                                                                                                                                                                                     [ 62%]
handler_misc.py::contract_check_asserts_work PASSED                                                                                                                                                                                                  [ 68%]
handler_read.py::contract_read_without_create PASSED                                                                                                                                                                                                 [ 75%]
handler_update.py::contract_update_read_success PASSED                                                                                                                                                                                               [ 81%]
handler_update.py::contract_update_list_success PASSED                                                                                                                                                                                               [ 87%]
handler_update_invalid.py::contract_update_create_only_property PASSED                                                                                                                                                                               [ 93%]
handler_update_invalid.py::contract_update_non_existent_resource PASSED                                                                                                                                                                              [100%]

============================================================================================================== 16 passed in 684.48s (0:11:24) ==============================================================================================================

```


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/rpdk/core/contract/resource_client.py [hunk 1]
```diff
@@ -186,6 +187,21 @@ def assert_write_only_property_does_not_exist(self, resource_model):
...
+    def get_metadata(self, metadata_key):
+        metadata = []
+        try:
+            properties = self._schema["properties"]
+        except KeyError:
+            return metadata
+
+        for prop in properties:
+            try:
+                if properties[prop][metadata_key] == "false":
```
Comment (prerna-p) [inline src/rpdk/core/contract/resource_client.py:199]: we are returning properties that have the metadata_key and a specific value. Maybe we could add a parameter that takes the value as well.

File: src/rpdk/core/contract/suite/handler_commons.py [hunk 2]
```diff
@@ -168,9 +168,22 @@ def test_input_equals_output(resource_client, input_model, output_model):
...
+            else:
+                assert (
+                    pruned_input_model[key] == pruned_output_model[key]
+                ), assertion_error_message
     except KeyError as e:
         raise AssertionError(assertion_error_message) from e
+
+
+def test_unordered_list_match(inputs, outputs):
+    assert len(inputs) == len(outputs)
```
Comment (prerna-p) [inline src/rpdk/core/contract/suite/handler_commons.py:185]: is it possible that output will have more values (set on service side) than input?

File: src/rpdk/core/contract/resource_client.py (hunk 1 above)
Comment (anshikg) [reply src/rpdk/core/contract/resource_client.py:199]: I am not sure I understand. Could you give me an example?

File: src/rpdk/core/contract/suite/handler_commons.py (hunk 2 above)
Comment (anshikg) [reply src/rpdk/core/contract/suite/handler_commons.py:185]: It is possible. Some resources have generated values and if so, they need to define the default values

**Comment by PatMyron** [comment]
Should `uniqueItems` be handled here as well? Only seeing [`insertionOrder`](https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/#insertionorder) so far

File: src/rpdk/core/contract/resource_client.py (hunk 1 above)
Comment (prerna-p) [reply src/rpdk/core/contract/resource_client.py:199]:
```python
get_metadata(self, metadata_key, value):
     .
     .
     .
      if properties[prop][metadata_key] == value:

File: src/rpdk/core/contract/suite/handler_commons.py (hunk 2 above)
Comment (prerna-p) [reply src/rpdk/core/contract/suite/handler_commons.py:185]: I think that will cause an issue here?

File: src/rpdk/core/contract/resource_client.py [hunk 3]
```diff
@@ -186,6 +187,21 @@ def assert_write_only_property_does_not_exist(self, resource_model):
...
+                    metadata.append(prop)
+            except KeyError:
+                continue
```
Comment (ammokhov) [inline src/rpdk/core/contract/resource_client.py:202]: minor: i'd keep `pass` to be consistent with other parts of implementation

File: src/rpdk/core/contract/resource_client.py (hunk 1 above)
Comment (ammokhov) [reply src/rpdk/core/contract/resource_client.py:199]: maybe something like:
```python
metadata = set()

metadata.add(
[
    prop for prop in properties.keys() if "insertionOrder" in properties[prop] and properties[prop]["insertionOrder"] == "false"
])
```

File: src/rpdk/core/contract/resource_client.py [hunk 4]
```diff
@@ -166,6 +166,7 @@ def _update_schema(self, schema):
         self.read_only_paths = self._properties_to_paths("readOnlyProperties")
         self.write_only_paths = self._properties_to_paths("writeOnlyProperties")
         self.create_only_paths = self._properties_to_paths("createOnlyProperties")
+        self.insertion_order = self.get_metadata("insertionOrder")
```
Comment (ammokhov) [inline src/rpdk/core/contract/resource_client.py:169]: can we name it more explicit, something like `properties_without_insertion_order`

File: src/rpdk/core/contract/resource_client.py [hunk 5]
```diff
@@ -186,6 +187,21 @@ def assert_write_only_property_does_not_exist(self, resource_model):
             ), "The model MUST NOT return properties defined as \
                 writeOnlyProperties in the resource schema"
 
+    def get_metadata(self, metadata_key):
```
Comment (ammokhov) [inline src/rpdk/core/contract/resource_client.py:190]: if you want to reuse this method, then would be great not to hardcode `false` but use it as a arg with default (would be great for variation)

**Review by ammokhov — CHANGES_REQUESTED** [review]

File: src/rpdk/core/contract/suite/handler_commons.py (hunk 2 above)
Comment (anshikg) [reply src/rpdk/core/contract/suite/handler_commons.py:185]: could you please elaborate, I am probably missing something

File: src/rpdk/core/contract/resource_client.py (hunk 1 above)
Comment (anshikg) [reply src/rpdk/core/contract/resource_client.py:199]: > ```python
> get_metadata(self, metadata_key, value):
>      .
>      .
>      .
>       if properties[prop][metadata_key] == value:
> ```

This is a valid point. This is useful in cases we have multiple metadata to check. In this case, I think we only care about insertion order. Might be worth removing the metadat_key too

Comment (anshikg) [reply src/rpdk/core/contract/resource_client.py:199]: > maybe something like:
> 
> ```python
> metadata = set()
> 
> metadata.add(
> [
>     prop for prop in properties.keys() if "insertionOrder" in properties[prop] and properties[prop]["insertionOrder"] == "false"
> ])
> ```

will make the change

File: src/rpdk/core/contract/resource_client.py (hunk 5 above)
Comment (anshikg) [reply src/rpdk/core/contract/resource_client.py:190]: valid point. Not sure we will be reusing this, I was thinking of hardcoding the metadat_key too

**Comment by anshikg** [comment]
> Should `uniqueItems` be handled here as well? Only seeing [`insertionOrder`](https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/#insertionorder) so far

hey Pat, I was initially going to add unique metadata check but, I realized the schema validation already takes care of this.I will update the commit message when merging

**Review by ammokhov — APPROVED** [review]

File: src/rpdk/core/contract/suite/handler_commons.py [hunk 6]
```diff
@@ -168,9 +168,25 @@ def test_input_equals_output(resource_client, input_model, output_model):
...
```
Comment (prerna-p) [inline src/rpdk/core/contract/suite/handler_commons.py:188]: Minor comment: I'm not clear on why this check is needed?

**Review by prerna-p — APPROVED** [review]

File: src/rpdk/core/contract/suite/handler_commons.py (hunk 6 above)
Comment (anshikg) [reply src/rpdk/core/contract/suite/handler_commons.py:188]: insertion order only care about the list being in-order or unordered. The list should still be of the same lenght

**Pull request merged**